### ISSUE TITLE
Public key auth fails if with signed public keys auth

### ISSUE THREAD

**Issue by lsiudut** [ronf/asyncssh#517]
First of all, I'm not even close to being a crypto expert so I might have misunderstood certain concepts. I'm pretty sure that `asyncssh` doesn't handle authentication with signed public keys properly.

To start with, after certain servers got upated to new OpenSSH (w/ default config) an authentication failures started to occur. After enabling full debug output and comparing official client verbose logs and asyncssh verbose logs I noticed that they use different methods to authenticate:

* official clients chose `rsa-sha2-512-cert-v01@openssh.com` method - to my understanding it picked a private key and a certificate to sign it (`id_rsa` and `id_rsa-cert.pub`, logs refer to it as `RSA-CERT SHA256`), signed the PK w/ the certificate and send it for authentication
* asyncssh is doing something similar, also reading both - PK and the cert - but the resulting type is `ssh-rsa-cert-v01@openssh.com`, which is not accepted by the server

After inspecting the `id_rsa-cert.pub` file I figured that its type is `ssh-rsa-cert-v01@openssh.com` and this is what `asyncssh` reads and stores after loading and signing the PK. I do suspect though that signed key should have different type which is `rsa-sha2-512-cert-v01@openssh.com` - after patching `decode_ssh_public_key` to override the algorithm to `rsa-sha2-512-cert-v01@openssh.com` it works just fine. To illustrate what I mean:

```
def decode_ssh_public_key(data: bytes) -> SSHKey:
    """Decode a packetized SSH public key"""

    try:
        packet = SSHPacket(data)
        alg = packet.get_string()

        # please don't  judge me
        if alg == b'ssh-rsa-cert-v01@openssh.com':
            alg = b'rsa-sha2-256-cert-v01@openssh.com'

        handler = _public_key_alg_map.get(alg)

        if handler:
            key_params = handler.decode_ssh_public(packet)
            packet.check_end()

            key = handler.make_public(key_params)
            key.algorithm = alg
            return key
        else:
            raise KeyImportError('Unknown key algorithm: %s' %
                                 alg.decode('ascii', errors='replace'))
    except PacketDecodeError:
        raise KeyImportError('Invalid public key') from None
```

I can't even count how many good practices does the above break, but it nicely proves the point. The result is that `alg` gets changed to the one that server expects, but certificate type remains unchanged and it's being properly processed deeper in the stack.

I suspect that it may be also something related to processing pk (https://github.com/ronf/asyncssh/issues/428), but once again - it's beyond my expertise to discuss cryptographic flows :).

**Comment by ronf** [comment]
Thanks for the report!

I think I've potentially been able to reproduce this here. Could you give the following patch a try?

```patch
diff --git a/asyncssh/public_key.py b/asyncssh/public_key.py
index 36c2faf..f632d38 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -2189,6 +2189,8 @@ class SSHKeyPair:

             cert = cast('SSHX509CertificateChain', self._cert)
             self.public_data = cert.adjust_public_data(sig_algorithm)
+        else:
+            self.algorithm = sig_algorithm + b'-cert-v01@openssh.com'

     def sign(self, data: bytes) -> bytes:
         """Sign a block of data with this private key"""
```

This patch is designed to be applied to the latest code in the "develop" branch, but I suspect it should apply pretty well to any reasonably recent version.

This adjusts the signature algorithm reported for a certificate, in addition to changing the actual signing algorithm itself to match one of the negotiated values. Previously, it would select a proper signature algorithm, but it left the certificate algorithm unchanged, and I think that caused OpenSSH to reject the key, since the default algorithm of 'ssh-rsa-cert-v01@openssh.com' is now marked by OpenSSH as an unsupported algorithm.

Without this patch, I suspect things might work if you are able to create a certificate with one of the SHA-2 algorithms (rsa-sha2-256-cert-v01@openssh.com or rsa-sha2-512-cert-v01@openssh.com). However, doing this automatically based on the chosen signature algorithm is much better. That's what the above patch should do.

A bit more cleanup may be needed here to properly handle the case of sk-ecdsa and sk-ssh-ed25519 keys, but first I want to confirm if this works for you in the RSA key case.

**Comment by ronf** [comment]
Here's a version of the above change which works properly with FIDO/U2F (sk) keys:

```patch
diff --git a/asyncssh/public_key.py b/asyncssh/public_key.py
index 36c2faf..e291c48 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -2189,6 +2189,11 @@ class SSHKeyPair:

             cert = cast('SSHX509CertificateChain', self._cert)
             self.public_data = cert.adjust_public_data(sig_algorithm)
+        else:
+            if sig_algorithm.endswith(b'@openssh.com'):
+                sig_algorithm = sig_algorithm[:-12]
+
+            self.algorithm = sig_algorithm + b'-cert-v01@openssh.com'

     def sign(self, data: bytes) -> bytes:
         """Sign a block of data with this private key"""
```

**Comment by lsiudut** [comment]
I can confirm it works! Some good learning came out from this for me. Thanks! :)

**Comment by ronf** [comment]
Terrific - thanks for the confirmation, and for reporting this! This is now checked into the "develop" branch as commit 5a22b1f and will be included in the next release.

To provide a bit more info, https://www.rfc-editor.org/rfc/rfc4252#section-7 gives details of the user auth request message involved, and the field I changed above is supported to the algorithm name of the public key or certificate being used to do the signing. However, when SHA-2 signature methods were added to RSA keys in https://www.ietf.org/rfc/rfc8332.txt, the key type didn't actually change, and so arguably the certificate type should also not be changing. In fact, at the time the certificate is created and signed, it can't KNOW what signature algorithm might be used with it, so you should always see the "ssh-rsa-cert-v01@openssh.com" algorithm name in the cert itself. So, I think it's actually wrong for OpenSSH to reject such a key type in cases where the actual signature is SHA-2.

If you look deeper into the signature blob (as described in https://www.rfc-editor.org/rfc/rfc4253#section-6.6), it has its own separate algorithm name in it, where you'd see something like "rsa-sha2-256" or "rsa-sha2-512" when those were used. So, even without changing the public key/cert type in the user auth message, the recipient has enough information to know exactly which signature algorithm was used, and OpenSSH could have done the enforcement based on that.

That said, I'm happy to be more interoperable here, and I really appreciate you filing this with enough detail for me to reproduce it... Thanks again!

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.13.0.
